### PR TITLE
Deposit contract and genesis handling

### PR DESCRIPTION
**Pull request by mpetrunic**
Creating of beacon chain genesis state is still WIP as I'm missing some new types and functions but it should be solved with #318

### CODE REVIEW

File: packages/lodestar/src/chain/chain.ts [hunk 1]
```diff
@@ -222,4 +190,54 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
...
+    const genesisState = initializeBeaconStateFromEth1 (
+      this.config,
+      Buffer.from(eth1Block.hash.replace("0x", ""), "hex"),
+      eth1Block.timestamp,
+      depositsWithProof
+    );
+    if(!isValidGenesisState(this.config, genesisState)) {
+      this.logger.info(`Eth1 block ${eth1Block.hash} is NOT forming valid genesis state`);
+    }
+    this.logger.info(`Initializing beacon chain with eth1 block ${eth1Block.hash}`);
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts:240]: Should I extract bottom part to public `initBeaconChain` method to get ahead of starting beaconchain from fake genesis state

File: packages/lodestar/src/chain/chain.ts [hunk 2]
```diff
@@ -7,51 +7,55 @@ import BN from "bn.js";
...
   private logger: ILogger;
 
-  public constructor(opts, {config, db, eth1, logger}: { config: IBeaconConfig; db: IBeaconDb; eth1: IEth1Notifier; logger: ILogger }) {
+  public constructor(opts, {config, db, eth1, opPool, logger}: IBeaconChainModules) {
     super();
     this.chain = opts.chain;
     this.config = config;
     this.db = db;
     this.eth1 = eth1;
+    this.opPool = opPool;
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/chain.ts:58]: I think this creates a circular dependency

**Review by wemeetagain — CHANGES_REQUESTED** [review]
LGTM other than circular dependency

File: packages/lodestar/src/chain/chain.ts (hunk 1 above)
Comment (wemeetagain) [reply packages/lodestar/src/chain/chain.ts:240]: Just was about to leave a comment to do that xD

File: packages/lodestar/src/chain/chain.ts (hunk 2 above)
Comment (mpetrunic) [reply packages/lodestar/src/chain/chain.ts:58]: Oh, damn. Any idea how to proceed. All my ideas are bad in some way: 
- call opPool.removeOperations directly 
- take chain module as param in start method of opPool
- fetch deposits directly from database in chain module

I don't like any of it. Maybe to pass some OpPool in a Promise :/

Comment (GregTheGreek) [reply packages/lodestar/src/chain/chain.ts:58]: How is this circular?

Comment (mpetrunic) [reply packages/lodestar/src/chain/chain.ts:58]: Eth1 was depending on OpPool, OpPool was depending on chain and chain was depending on opPool and eth1. No way to initialize them. Fixed it in latest commit

File: packages/lodestar/src/chain/chain.ts [hunk 3]
```diff
@@ -213,28 +237,8 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
     if(!isValidGenesisState(this.config, genesisState)) {
       this.logger.info(`Eth1 block ${eth1Block.hash} is NOT forming valid genesis state`);
     }
+    await this.initializeBeaconChain(genesisState, merkleTree);
```
Comment (GregTheGreek) [inline packages/lodestar/src/chain/chain.ts]: semantic, but the log is incorrect, either change the log to `Initialized` as its completed or swap line 240 and 241

**Review by wemeetagain — CHANGES_REQUESTED** [review]
can you remove debug and coverage files?

**Comment by GregTheGreek** [comment]
Spec Tests really don't like this pr, otherwise LGTM

**Review by wemeetagain — APPROVED** [review]

**Comment by wemeetagain** [comment]
LGTM, FYI, I'm working off these changes to bring the types / state transition up to spec

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**